File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        
        def valid(i, j):
            return not (i<0 or j<0 or i>=m or j>=n or grid[i][j]!=1)
        
        q=[]
        fresh, m , n = 0, len(grid), len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j]==2:
                    q.append([i, j, 0])
                if grid[i][j]==1:
                    fresh+=1
        time=0
        while q and fresh:
            i, j, t = q.pop(0)
            if grid[i][j]==1:
                fresh-=1
            grid[i][j]=3
            time=max(time, t)
            t+=1
            if valid(i+1, j):
                q.append([i+1, j, t])
            if valid(i, j+1):
                q.append([i, j+1, t])
            if valid(i-1, j):
                q.append([i-1, j, t])
            if valid(i, j-1):
                q.append([i, j-1, t])
        
        return time if fresh==0 else -1

File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        
        int m = grid.length;
        int n = grid[0].length;
        int fresh = 0;
        int[][] dir = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        Queue<int[]> q = new LinkedList<>();
        
        for(int i=0; i<m; ++i){
            for(int j=0; j<n; ++j){
                if(grid[i][j] == 2)
                    q.offer(new int[]{i, j});
                else if(grid[i][j] == 1)
                    ++fresh;
            }
        }
        
        // BFS
        // fresh > 0 so that we dont have to do return time-1;
        int time = 0;
        while(!q.isEmpty() && fresh>0){
            int l = q.size();
            for(int a=0; a<l; ++a){                
                int i = q.peek()[0];
                int j = q.peek()[1];
                q.poll();
                
                for(int[] d: dir){
                    int x = i+d[0];
                    int y = j+d[1];
                    if(x<0 || x>=m || y<0 || y>=n || grid[x][y] != 1)
                        continue;
                    grid[x][y] = 2;
                    --fresh;
                    q.offer(new int[]{x, y});
                }
            }
            ++time;
        }
        
        return fresh==0 ? time : -1;
    }
}


// class Solution {
//     int m;
//     int n;
//     public int orangesRotting(int[][] grid) {
//         m = grid.length;
//         n = grid[0].length;
//         int[][] dir = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
//         Queue<int[]> q = new LinkedList<>();
        
//         for(int i=0; i<m; ++i){
//             for(int j=0; j<n; ++j){
//                 if(grid[i][j] == 2)
//                     q.offer(new int[]{i, j});
//             }
//         }
        
//         int time = 0;
//         // BFS
//         while(!q.isEmpty()){
//             int l = q.size();
//             for(int a=0; a<l; ++a){                
//                 int i = q.peek()[0];
//                 int j = q.peek()[1];
//                 q.poll();
                
//                 for(int[] d: dir){
//                     int x = i+d[0];
//                     int y = j+d[1];
//                     if(x<0 || x>=m || y<0 || y>=n || grid[x][y] != 1)
//                         continue;
//                     grid[x][y] = 2;
//                     q.offer(new int[]{x, y});
//                 }
//             }
//             ++time;
//         }
//         for(int i=0; i<m; ++i){
//             for(int j=0; j<n; ++j){
//                 if(grid[i][j]==1)
//                     return -1;
//             }
//         }
        
//         // Math.max because testcase [[0]]
//         return Math.max(time-1, 0);
//     }
// }